src/aarch64/trampoline.S

```diff
@@ -0,0 +1,44 @@
+#ifdef __arm64__
+
+#define LIBFFI_ASM
+#include <fficonfig.h>
+#include <ffi.h>
+#include <ffi_cfi.h>
+#include "internal.h"
+
+#ifdef FFI_EXEC_TRAMPOLINE_TABLE
+
+#ifdef __MACH__
+#include <mach/machine/vm_param.h>
+#endif
+
+#ifdef HAVE_MACHINE_ASM_H
+# include <machine/asm.h>
+#else
+# ifdef __USER_LABEL_PREFIX__
+#  define CONCAT1(a, b) CONCAT2(a, b)
+#  define CONCAT2(a, b) a ## b
+#  define CNAME(x) CONCAT1 (__USER_LABEL_PREFIX__, x)
+# else
+#  define CNAME(x) x
+# endif
+#endif
+
+.set page_max_size, PAGE_MAX_SIZE
+.align PAGE_MAX_SHIFT
+.text
+.globl CNAME(ffi_closure_trampoline_table_page)
+CNAME(ffi_closure_trampoline_table_page):
+    .rept PAGE_MAX_SIZE / FFI_TRAMPOLINE_SIZE
+#ifdef FFI_TRAMPOLINE_WHOLE_DYLIB
+    adr x16, -(2 * PAGE_MAX_SIZE)
+#else
+    adr x16, -PAGE_MAX_SIZE
+#endif
+    ldp x17, x16, [x16]
+    BRANCH_TO_REG x16
+	nop		/* each entry in the trampoline config page is 2*sizeof(void*) so the trampoline itself cannot be smaller that 16 bytes */
+    .endr
+
+#endif /* FFI_EXEC_TRAMPOLINE_TABLE */
+#endif /* __arm64 */
```

Comment:
Build system changes seem to be missing for this file.

Reply:
Yeah, the build system changes were specific to our Xcode project.  I'm not looking forward to figuring that out for autotools.

Reply:
Ah no worries, I can take a stab at it and post a suggestion here.

Reply:
Just pushed a rough draft [here](https://github.com/frida/libffi/tree/apple-arm64e).

Here's how I tested it:

### Generate a build environment

```sh
$ git clone https://github.com/frida/frida.git
$ cd frida
$ make build/frida-env-ios-arm64e.rc
```

> Piggybacking on the autotools legacy build environment in [Frida](https://frida.re/), which we still use for building a few dependencies that haven't yet moved to [Meson](https://mesonbuild.com/).

### Bootstrap and build for iOS/arm64e

```
$ . build/frida-env-ios-arm64e.rc
$ cd ~/src/libffi
$ ./autogen.sh
$ ./configure --prefix=/usr --enable-shared
$ make
$ make install DESTDIR="$(echo ~/Desktop/libffi-ios-arm64e)"
```

> --enable-shared is only needed because Frida's $CONFIG_SITE defaults to enable_shared=no

### Build for macOS

From a freshly initialized shell:

```sh
$ . ~/src/frida/build/frida-env-macos-x86_64.rc
$ ./configure --prefix=/usr --enable-shared
$ make
$ make install DESTDIR="$(echo ~/Desktop/libffi-macos-x86_64)"
```

Reply:
Cool, thanks for the help.  FWIW, I am *crazy* busy with WWDC work and taking care of kids at home, but I plan to circle back to this after WWDC, break out some of the smaller less offensive pieces into individual PRs, and then polish up the rest.